integrationtests/src/test/java/io/kroxylicious/proxy/filter/ApiVersionsMarkingFilterFactory.java

```diff
@@ -6,20 +6,54 @@
 
 package io.kroxylicious.proxy.filter;
 
-public class ApiVersionsMarkingFilterFactory implements FilterFactory<ApiVersionsMarkingFilter, Void> {
+import java.util.concurrent.CompletionStage;
+
+import org.apache.kafka.common.message.ApiVersionsResponseData;
+import org.apache.kafka.common.message.RequestHeaderData;
+import org.apache.kafka.common.protocol.ApiKeys;
+import org.apache.kafka.common.protocol.ApiMessage;
+import org.apache.kafka.common.protocol.types.RawTaggedField;
+
+import io.kroxylicious.proxy.ApiVersionsService;
+import io.kroxylicious.proxy.plugin.PluginConfigType;
+
+import static java.nio.charset.StandardCharsets.UTF_8;
+
+@PluginConfigType(Void.class)
+public class ApiVersionsMarkingFilterFactory implements FilterFactory<Void, Void> {
```

Comment:
This comment is not specific to this instance but I wanted to talk about it in context. 

For me the core thing we want Filter Authors to be focused on is the business logic required to implement their goal. Working out what ever observations or transformations of the message are required. 

By arranging things so that FilterFactory is the primary thing that needs to be implemented we are changing the focus of filter authors. Say they are wanting to test drive their implementation starting with  FilterFactory just gets in the way of the tests they want to write. 

As we want to use ServiceLoaders to find contributed filters and Service Providers have to be a public top level class we can't make the FilterFactory an inner class. 
However, I wonder if we can get the best of both worlds. We make marker interface `Filter` the service that we load. Filters can be annotated with the annotations proposed in this PR `@PluginConfig` and `@PluginConfigType()` and optionally `@FilterFactory`. I'm contending that the FilterFactory annotation is optional as I imagine most filter authors are just going to have a no arg constructor or just pass their config directly to the filter instance. To that end we can then default the filter factory to choice to something like the following:

```java
import java.lang.reflect.Constructor;

import io.kroxylicious.proxy.filter.FilterContext;

public class ArbitaryFilterFactory<C, I, F extends Filter> implements io.kroxylicious.proxy.filter.FilterFactory<C, I> {

    private final Class<F> filterClass;

    public ArbitaryFilterFactory(Class<F> filterClass) {
        this.filterClass = filterClass;
    }

    @SuppressWarnings("unchecked")
    @Override
    public I initialize(FilterFactoryContext context, C config) throws FilterFactoryException {
        return (I) config;
    }

    @Override
    @NonNull
    public Filter createFilter(FilterFactoryContext context, I initializationData) {
        try {
            if (initializationData == null) {
                return filterClass.getConstructor().newInstance();
            }
            else {
                final Constructor<F> constructor = filterClass.getConstructor(initializationData.getClass());
                return constructor
                        .newInstance(initializationData);
            }
        }
        catch (InstantiationException | IllegalAccessException | InvocationTargetException | NoSuchMethodException e) {
            throw new RuntimeException(e);
        }
    }
}
```

If the filter author uses the filterFactory annotation we would then need to load that factory (presumably the current service loader mechanism) and delegate to it.

I haven't fully grokked the entirety of the PR as I've been tripping over this thought.

Reply:
> By arranging things so that FilterFactory is the primary thing that needs to be implemented we are changing the focus of filter authors. 

Just to be clear,  although I did change all the existing filters, nothing in this PR _requires_ the filter to be an inner class of the factory. So it's entirely up to us to agree how we want to explain it to authors, and it's up to authors how they want to write it.

FWIW I've been backing out this change this afternoon. Not because I've changed my mind, but just to make the PR smaller. 

> We make marker interface Filter the service that we load.

I don't think we should do that. The reason is that instantiating a service (or equivalently, reflection) will always be slower than `new`, and we need to instantiate `Filter` instances on every connection, so it's a relatively hot path.  By indirecting via `FilterFactory` we get to instantiate the service _once_ during startup. Then on the hot path we're just calling the factory (`invokevirtual`) which is invoking `new`.

Reply:
> We make marker interface Filter the service that we load. Filters can be annotated with the annotations proposed in this PR @PluginConfig and @PluginConfigType() and optionally @FilterFactory.

There's a difference between when we interrogate the ServiceLoader for the available providers and when we call `Provider.get()` to get an instance of the provided service. We can (and do) do the interrogation at start up. But that doesn't avoid the costs of the `get()` call, which happen when we're setting up the filter chain for a specific channel. That `get()` call is costly, because it uses reflection. 

So if `Filters` are services then means we're doing this reflection on the the hot path. This is basically why we have the `FilterFactory` in the first place. See the "Designing services " in the [javadoc for ServiceLoader](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/ServiceLoader.html)